addons/website_sale/controllers/main.py

```diff
@@ -909,6 +915,86 @@ def address(self, **kw):
         render_values.update(self._get_country_related_render_values(kw, render_values))
         return request.render("website_sale.address", render_values)
 
+    @http.route(
+        ['/shop/express'], type='json', methods=['POST'],
+        auth="public", website=True, sitemap=False,
+    )
+    def process_express_checkout(self, **post):
+        """ TODO VCR: do a proper docstring
+        TODO VCR: FIX ADDRESS CHANGING
+        """
+        def _create_or_edit_billing_partner(order, billing_address, edit=False):
+            mode = ('edit' if edit else 'new', 'billing')
+            billing_pre_values = self.values_preprocess(order, mode, billing_address)
+            billing_errors, billing_error_msg = self.checkout_form_validate(
+                mode, billing_address, billing_pre_values
+            )
+            billing_new_values, billing_errors, billing_error_msg = self.values_postprocess(
+                order, mode, billing_pre_values, billing_errors, billing_error_msg
+            )
+            billing_partner_id = self._checkout_form_save(
+                mode, billing_new_values, billing_address
+            )
+            if not billing_errors:
+                # We need to set billing partner first to be able to save shipping partner.
+                order.partner_id = billing_partner_id
+                # This is the *only* thing that the front end user will see/edit anyway when
+                # choosing billing address.
+                order.partner_invoice_id = billing_partner_id
+                order.message_partner_ids = [Command.set([billing_partner_id])]
+            else:
+                billing_errors['error_message'] = billing_error_msg
+                raise ValidationError(billing_errors)
+
+        order = request.website.sale_get_order()
+
+        partner_id = int(post['partner_id'])
+
+        # If the partner exists, we use the address we already know. If he doesn't have a complete
+        # address yet, we use the address from the form.
+        if partner_id > 0:
+            partner_sudo = request.env['res.partner'].browse(partner_id).sudo()
+            if not all(partner_sudo[key] for key in ('city', 'country_id', 'street')):
+                # Complete billing address.
+                self.include_country_and_state_in_address(post['billing_address'])
+                post['billing_address']['partner_id'] = partner_id  # Specifically for the 'edit' case.
+                del post['billing_address']['email']  # Don't replace the existing email.
+                _create_or_edit_billing_partner(order, post['billing_address'], edit=True)
+            else:
+                if not partner_sudo.phone:
+                    partner_sudo.phone = post['billing_address']['phone']
```

Comment:
Since the route is public and this is `sudo`-ed, would it not be possible for anyone to just update the information of arbitrary partners here?